src/gdal/ogr_file_format.cpp

```diff
@@ -1408,5 +1383,593 @@ LatLon OgrFileImport::calcAverageLatLon(OGRDataSourceH data_source)
+		if (point_layer != nullptr)
+		{
+			addPointsToLayer(point_layer,
+			                            [](const Object* object){ return object->getSymbol()->getType() == Symbol::Point; });
+			addTextToLayer(point_layer,
+			                            [](const Object* object){ return object->getSymbol()->getType() == Symbol::Text; });
+		}
+		else
+		{
+			addWarning(tr("Unable to export point and text objects"));
+		}
+
+		auto line_layer = createLayer(QString::fromLatin1("%1_lines").arg(output_file_info.baseName()).toLatin1(), wkbLineString);
+		if (line_layer != nullptr)
+			addLinesToLayer(line_layer,
+			                [](const Object* object){ return object->getSymbol()->getType() == Symbol::Line; });
+		else
+			addWarning(tr("Unable to export line objects"));
+
+		auto area_layer = createLayer(QString::fromLatin1("%1_areas").arg(output_file_info.baseName()).toLatin1(), wkbPolygon);
+		if (area_layer != nullptr)
+			addAreasToLayer(area_layer,
+			                [](const Object* object){ return object->getSymbol()->getType() == Symbol::Area ||
+				        object->getSymbol()->getType() == Symbol::Combined; });
+		else
+			addWarning(tr("Unable to export area objects"));
+	}
+
+	// Close the dataset to flush all data
+	po_ds.reset();
+
+	// Copy files to IODevice
+	vsiCopyFiles();
+}
+
+void OgrFileExport::vsiCopyFiles()
+{
+	// Go through our virtual filesystem
+	std::unique_ptr<char*, decltype(&CSLDestroy)> names(VSIReadDir(vsimem_path), &CSLDestroy);
+	for (auto i = 0; names.get()[i] != NULL; i++)
+	{
+		VSIStatBufL s_stat;
+		QString vsi_full_path = QString::fromLatin1(vsimem_path).append(QString::fromLatin1(names.get()[i]));
+		if (VSIStatExL(vsi_full_path.toLatin1(), &s_stat, VSI_STAT_NATURE_FLAG) == 0)
+		{
+			if (VSI_ISDIR(s_stat.st_mode))
+			{
+				// VSIReadDir includes the current path in output
+				if (vsi_full_path.compare(QString::fromLatin1(vsimem_path)))
+					addWarning(tr("Found unknown directory %1.  Please report this as a bug.").append(vsi_full_path));
+			} else {
+				if (output_file_info.fileName().compare(QString::fromLatin1(names.get()[i])) == 0)
+				{
+					// Write to IODevice
+					vsi_l_offset length;
+					std::unique_ptr<GByte,  decltype(&std::free)> pointer(
+					            VSIGetMemFileBuffer(vsi_full_path.toLatin1(), &length, TRUE), &std::free);
+					if (pointer.get() == nullptr)
+						throw FileFormatException(tr("Failed to read from temporary file in memory"));
+					io_device->write(reinterpret_cast<const char*>(pointer.get()), length);
+				}
+				else
+				{
+					// Handle any "extra" files, eg dbf, shx for shapefiles
+					// VSIRename() is only for copying within same folder and can't be used here
+					QFile out_file(output_file_info.absolutePath()
+					               .append(QString::fromLatin1("/")).append(QString::fromLatin1(names.get()[i])));
+					if (!out_file.open(QIODevice::WriteOnly))
```

Comment:
Should an exception or warning be added if the attempted output file already exists?